beps/0002-dynamic-frontend-plugins/README.md

```diff
@@ -154,12 +154,386 @@ Each plugin is required to provide manifest file (metadata) in predefined format
 
 Plugin discovery is a pre-requisite for Plugin registry. This should be responsible for scanning for available plugins and generating/modifying the plugin registry to always keep it up to date.
 
-## Design Details
+### CLI
+
+The following sections contains proposal of new CLI scripts to support the dynamic frontend plugins.
+
+#### Static and Dynamic Plugin Build Differences
+
+The major difference between static and dynamic plugins lies in their dependencies and how they are "installed" into a Backstage instance and whether they are bundled into final JS output during the build.
+
+While static plugins can define their dependencies using the dependencies field inside `package.json`, dynamic (module federation-based) plugins can't.
+
+A dynamic plugin requires all dependencies to be present in the build output. Module sharing is then used to optimize the "output size" at runtime. If a shared package is not found in the browser scope, or if the version range does not match the requirement, it falls back and loads additional JS assets from the "private" build and adds them to the scope.
+
+Webpack is also not the best tool for building traditional NPM packages. Dependencies can be externalized, but Webpack brings additional code into the bundles. Using Webpack external dependencies also conflicts with module federation shared modules. Externals will not be included even as a shared module fallback. As a result, such shared module can't be used because the dependency will not be defined in the init container.
+
+If the vision is to still support "static" plugins installed via NPM, it makes sense to keep the static and dynamic plugin builds separate and build two different bundles. Both can still be published as a single NPM package (not recommended), but to achieve optimal performance for both, there cannot be a shared output.
```

Comment:
Got a bit nerd-sniped here thinking: what if we could do exactly this? Having a single build that is compatible with both dynamic and build-time bundling? When bundling the app you'd simply ignore all of the bundled transitive dependencies. As you mention I think it would require some very ugly solution, but it does have me intrigued 🤔 😁

Reply:
I was trying to think of a solution that would allow us exactly this. The problem is the dependencies though :face_exhaling: .

> what if we could do exactly this?

I think it can be done. But it is not trivial. We might have to dig into some custom webpack plugins or even worse. I don't think its worth the effort **for now**. I'd consider this as a follow up.

The fact that the static plugins are supposed to be your traditional NPM packages is the problem here. Webpack is a must have for module federation. We have to bundle dependencies. Which is exactly what you don't want for NPM dependencies. Externals can't be used because it messes up the shared scopes. This problem does not exist in a world where the plugins do not depend on NPM ecosystem.

I think (maybe it's just a hope) that eventually, dynamic plugins will be the default way of using backstage plugins. And instead of NPM there will be an ecosystem of CNDs (like RHDH, roadiehq, Spotify) that will distribute their dynamic plugins.